src/sage/combinat/permutation.py

```diff
@@ -5383,6 +5383,253 @@ def shifted_shuffle(self, other):
         return self.shifted_concatenation(other, "right").\
         right_permutohedron_interval(self.shifted_concatenation(other, "left"))
 
+    def nth_roots(self, n):
+        r"""
+        Return all n-th roots of ``self`` (as a generator).
+
+        An n-th root of the permutation ``self`` is a permutation `\gamma` such that `\gamma^n == self`.
+
+        Note that the number of n-th roots only depend on the cycle type of ``self``.
+
+        EXAMPLES::
+
+            sage: Sigma = Permutations(5).identity()
+            sage: list(Sigma.nth_roots(3))
+            [[1, 4, 3, 5, 2], [1, 5, 3, 2, 4], [1, 2, 4, 5, 3], [1, 2, 5, 3, 4], [4, 2, 3, 5, 1], [5, 2, 3, 1, 4], [3, 2, 5, 4, 1],
+             [5, 2, 1, 4, 3], [2, 5, 3, 4, 1], [5, 1, 3, 4, 2], [2, 3, 1, 4, 5], [3, 1, 2, 4, 5], [2, 4, 3, 1, 5], [4, 1, 3, 2, 5],
+             [3, 2, 4, 1, 5], [4, 2, 1, 3, 5], [1, 3, 4, 2, 5], [1, 4, 2, 3, 5], [1, 3, 5, 4, 2], [1, 5, 2, 4, 3], [1, 2, 3, 4, 5]]
+
+            sage: Sigma = Permutation('(1, 3)')
+            sage: list(Sigma.nth_roots(2))
+            []
+
+        For n >= 6, this algorithm begins to be more efficient than naive search
+        (look at all permutations and test their n-th power).
+
+        .. SEEALSO::
+
+            * :meth:`has_nth_root`
+            * :meth:`number_of_nth_roots`
+
+        TESTS:
+
+        We compute the number of square roots of the identity (i.e. involutions in `S_n`, :oeis:`A000085`)::
+
+            sage: [len(list(Permutations(n).identity().nth_roots(2))) for n in range(2,8)]
+            [2, 4, 10, 26, 76, 232]
+
+            sage: list(Permutation('(1)').nth_roots(2))
+            [[1]]
+
+            sage: list(Permutation('').nth_roots(2))
+            [[]]
+
+            sage: Sigma = Permutations(6).random_element()
+            sage: list(Sigma.nth_roots(1)) == [Sigma]
+            True
+
+            sage: list(Permutations(4).identity().nth_roots(-1))
+            Traceback (most recent call last):
+            ...
+            ValueError: n must be at least 1
+        """
+        from sage.combinat.partition import Partitions
+        from sage.combinat.set_partition import SetPartitions
+        from itertools import product
+        from sage.arith.misc import divisors, gcd
+
+        def merging_cycles(list_of_cycles):
+            """
+            Generate all l-cycles such that its n-th power is the product
+            of cycles in 'cycles' (which contains gcd(l, n) cycles of length l/gcd(l, n))
```

Comment:
should there be backticks around "l", "n", "gcd(l, n)" and "1/gcd(l, n)"?  Also, possibly change "l" to "k" in the docstring (3-times)

Reply:
dostrings of inner auxiliary functions are never displayed, except when looking at the code. No need to format them